Refactor/dashboard index

eujiaxin opened the pull request:
## Changes made
Previously, translations are handled from data's side, such that BE returns `{'en': ..., 'bm': ...}`, and FE parses accordingly. i18n is now used to handle this with `dashboards.json` and `agencies.json`. The data (BE response) has been refactored to only include necessary key values, e.g. dashboard and agency key, and FE will get the appropriate translated string values via `useTranslation`.
1. Data fetching format & i18n namespaces for `getStaticProps()` in both `pages/dashboard` & `pages/dashboard/agency` are updated 
2. Dashboard component is updated following new fetched data format:
    - dedicated `dashboards` and `agencies` namespaces are used, e.g. `dashboards` give name and description of dashboards, `agencies` provide full and abbreviated form of agency names.
    - popular dashboards (section 1) are hidden when users start to search.
    - ** routing to the correct dashboard depends on the defined dashboard key from data - must make sure that parquet data dashboard value == i18n dashboard key == FE pages/dashboard/[dashboard-name]
3. `AgencyIcon: FunctionComponent<AgencyIconProps> = ({ agency })` - the agency keys are standardised to be consistent with keys in [i18n/[lang]/agencies.json](https://github.com/data-gov-my/datagovmy-meta/blob/main/i18n/en-GB/agencies.json), e.g. `"bnm": BNMIcon`

## Considerations
* Perhaps we should completely extract out `common:agency.[key]` to `agencies:[key]` separately for consistent agency naming. This mostly refers to refactoring AgencyBadge
* Similar to standardising agencies name, should each dashboard page refer to `dashboards` namespace for the dashboard name as well? so we don't have to update two places in i18n separately. however this might increase bundle size (?)
* Current way requires parquet data dashboard value == i18n dashboard key == FE pages/dashboard/[dashboard-name], maybe can define the routes separately in i18n, e.g. snippet below, however I do think that it may be a bit unnecessary, the current way is: 
```js
// BE response data format
item = {
          "name": "election-explorer",
          "agency": "spr",
          "views": 46744
      }
// redirect
<At href={`/dashboard/${item.name}`} locale={i18n.language}>
```
```js
// meta repo - i18n/[lang]/dashboards.json  

"categories": {
          "healthcare": "Healthcare", 
          "transport": "Transport",
...
        },
"dashboards": {
  "emergency-response": { // !!important!! dashboard name == route name
    "name": "Emergency Response",
    "description": "Real-time data for efficient emergency management."
    // "route": ... <- can do this if don't want to rely on key to redirect
  },
  "blood-donation": {
    "name": "Blood Donation",
    "description": "Track blood supply and encourage donations."
....
  },
  ```

irfancoder commented:
> * Perhaps we should completely extract out `common:agency.[key]` to `agencies:[key]` separately for consistent agency naming. This mostly refers to refactoring AgencyBadge

I can agree to this. But since most pages will be using the `agency` namespace, you need to add it together with where `common` is declared. (inside `withi18n(...)` function in `lib/decorator.ts`)


> * Similar to standardising agencies name, should each dashboard page refer to `dashboards` namespace for the dashboard name as well? so we don't have to update two places in i18n separately. however this might increase bundle size (?)


No, they only share title/name, but description and the rest arent the same, on top of importing extra i18n meant for `dashboard/index` on every dashboard. Keep the i18n scope narrow by the page or by the common traits


> * Current way requires parquet data dashboard value == i18n dashboard key == FE pages/dashboard/[dashboard-name], maybe can define the routes separately in i18n, e.g. snippet below, however I do think that it may be a bit unnecessary, the current way is:
> 
> ```js
> // BE response data format
> item = {
>           "name": "election-explorer",
>           "agency": "spr",
>           "views": 46744
>       }
> // redirect
> <At href={`/dashboard/${item.name}`} locale={i18n.language}>
> ```
> 
> ```js
> // meta repo - i18n/[lang]/dashboards.json  
> 
> "categories": {
>           "healthcare": "Healthcare", 
>           "transport": "Transport",
> ...
>         },
> "dashboards": {
>   "emergency-response": { // !!important!! dashboard name == route name
>     "name": "Emergency Response",
>     "description": "Real-time data for efficient emergency management."
>     // "route": ... <- can do this if don't want to rely on key to redirect
>   },
>   "blood-donation": {
>     "name": "Blood Donation",
>     "description": "Track blood supply and encourage donations."
> ....
>   },
> ```

No strong opinion on this. Both are viable ways to convey a route to FE. The only downside (not a big issue, just semantic) I can see is we will be using the `t("route")` function to refer to the route, when the wide use of it, is to translate i18n keys, correct? Is there any other benefit in doing this way?

eujiaxin commented:
> I can agree to this. But since most pages will be using the agency namespace, you need to add it together with where common is declared. (inside withi18n(...) function in lib/decorator.ts)

Updated to include `agencies` for all namespaces under withi18n(). Updated existing translation keys - `t(...)` that use `common:agency.[key]` to `agencies:[key].name` accordingly.

> No, they only share title/name, but description and the rest arent the same, on top of importing extra i18n meant for dashboard/index on every dashboard. Keep the i18n scope narrow by the page or by the common traits

Noted on this! :)

> No strong opinion on this. Both are viable ways to convey a route to FE. The only downside (not a big issue, just semantic) I can see is we will be using the t("route") function to refer to the route, when the wide use of it, is to translate i18n keys, correct? Is there any other benefit in doing this way?

 Defining the routes in i18n provides more flexibility in naming the routes, but I don't think its necessary, since they'll all follow the same convention - `/dashboard/[name]`, therefore this has not been changed. 


Additionally, `[agency]` in `dashboard/agency/[agency]` page is now cross-checked against valid agencies from the dropdown API, will return 404 if it doesn't exist in there. 
```ts
...
        get("/dropdown", { dashboard: "dashboards" }).then(res => {
          if (!res.data?.data.find((agency: string) => agency === params?.agency)) {
            throw `Invalid agency parameter: ${params?.agency}`;
          }
          return res.data;
        }),
```
